REGRESSION(301701@main): Web Inspector: crash in EventFiredCallback

yury-s opened the pull request:
#### 5e36299743f8d2543153358221fb8a46438a30c2
<pre>
REGRESSION(301701@main): Web Inspector: crash in EventFiredCallback
<a href="https://bugs.webkit.org/show_bug.cgi?id=301146">https://bugs.webkit.org/show_bug.cgi?id=301146</a>

Reviewed by Devin Rousso.

EventFiredCallback can live after its InspectorDOMAgent
has been destroyed. It can be collected in a later GC
cycle along with the DOM bindings at which point the reference
to the DOM agent is invalid. Use WeakPtr to avoid a crash
there.

* Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
* Source/WebCore/inspector/agents/InspectorDOMAgent.h:

Canonical link: <a href="https://commits.webkit.org/301844@main">https://commits.webkit.org/301844@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f2e1d5fb95862e76439f2c1baaa4f348ee8645c7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/127242 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/46890 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/38021 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/134306 "Built successfully") | [❌ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/78797 "Failed to checkout and rebase branch from PR 52696") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/129114 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/47500 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/55411 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/96835 "16 failures") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/59/builds/78797 "Failed to checkout and rebase branch from PR 52696") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/130190 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/38015 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/113960 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/77335 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/36898 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/32111 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/77686 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/107874 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/32454 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/136789 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/53901 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/41529 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/105354 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/54412 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/110311 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/105040 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26776 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/50562 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/29025 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/51464 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/53838 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/59925 "Built successfully") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/53070 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/56483 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/54831 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

dcrousso reviewed (approved):
this is i think fine as a short-term fix, but we should probably have logic in `InspectorDOMAgent::reset` that removes the listeners (and stops `m_mediaMetricsTimer`) as im not even sure GC will remove these since they're created entirely in C++

we also definitely dont need separate instances of `EventFiredCallback` for each node since it's the same handler for all of them (i.e. it could just be a single `RefPtr` member variable that's added and removed from all relevant nodes)

yury-s committed:
REGRESSION(301701@main): Web Inspector: crash in EventFiredCallback
https://bugs.webkit.org/show_bug.cgi?id=301146

Reviewed by Devin Rousso.

EventFiredCallback can live after its InspectorDOMAgent
has been destroyed. It can be collected in a later GC
cycle along with the DOM bindings at which point the reference
to the DOM agent is invalid. Use WeakPtr to avoid a crash
there.

* Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
* Source/WebCore/inspector/agents/InspectorDOMAgent.h:

Canonical link: https://commits.webkit.org/301844@main
